src/server/mgflask/mgflask/news.py

```diff
@@ -4,76 +4,63 @@
+from mgflask.models import Article
+from sqlalchemy import and_, or_
 
 bp = Blueprint('news', __name__, url_prefix='/news')
 
-api_keys = ['983d4d9ce3dc4f3badda1a1171eb548d', 'b5ad966ba07741858c365a83ed18a0bb']
-target_sources = ["bbc-news", "fox-news", "the-wall-street-journal", "national-review", "the-huffington-post", "the-hill", "cnn"]
-accepted_params= [ 'sources','qintitle','q','domains','excludeDomains', 'from', 'to', 'language', 'sortBy']
-accepted_params_int = ['page','pageSize']
-api_key_index_param = 'api_key_index'
+# Query parameters should be sent in json format in request body. Accepted parameters are values are defined below.
+# Each query parameter can contain a single or a list of accepted values. 
 
+# these columns are defined in model.py to match news.api parameters: https://newsapi.org/docs/endpoints/top-headlines
+# exact_columns need to match exactly
+EXACT_COLUMNS= [ 'id', 'author', 'source', 'url', 'urlToImage']
+# fuzzy_columns uses the LIKE operator to partically match
+FUZZY_COLUMNS = ['title', 'content', 'description']
+# for range columns the first two values will be interpreted as the lower and upper bouond and the rest discarded (if only one value then it will be the lower bound) 
+RANGE_COLUMNS = ['leftBias', 'right-Bias', 'publishedAt']
+# query on comments and article_ratings not yet implemented
 
-#get headlines from all target sources
-@bp.route('/headlines', methods=['GET'])
-def get_headlines_from_all():
-    topheadlines = {}
-    try:
-      api_key = get_api_key(request)
-      newsapi = NewsApiClient(api_key=api_key)
-      sources = ",".join(target_sources)
-      topheadlines = newsapi.get_top_headlines(sources=sources)
-    except (IndexError, ValueError)as e:
-      topheadlines['status'] = 'error'
-      topheadlines['message'] = str(e)
+# upper limit on the number of to return
+LIMIT_ARTICLE_PARAM = 'limit_articles'
 
-    return jsonify(topheadlines)
+#search among all articles
+@bp.route('/', methods=['GET'])
+def get_articles():
+    params = request.json
+    filters = []
 
-#get headlines from a designated target source
-@bp.route('/headlines/<news_source>', methods=['GET'])
-def get_headlines_from(news_source):
-    topheadlines = {}
-    try:
-      api_key = get_api_key(request)
-      newsapi = NewsApiClient(api_key=api_key)
-      topheadlines = newsapi.get_top_headlines(sources=news_source)
-    except (IndexError, ValueError) as e:
-      topheadlines['status'] = 'error'
-      topheadlines['message'] = str(e)
+    if params: 
+      for col in Article.__table__.columns:
+        param = params.get(col.key)
+        if col.key in EXACT_COLUMNS and param:
+          if isinstance(param, list):
+            filters.append(col.in_(param))
+          else:
+            filters.append( col==param )
+        elif col.key in FUZZY_COLUMNS and param:
+          if isinstance(param, list):
+            fuzzy_filters = [col.ilike("%"+each+"%") for each in param]
```

Comment:
could you explain what this does? im confused

Reply:
It's using the sqlalchemy Query API https://docs.sqlalchemy.org/en/14/orm/query.html. I took some ideas from https://stackoverflow.com/questions/3332991/sqlalchemy-filter-multiple-columns

Reply:
basically I am implementing exact matching on mostly columns with one-word content ('id', 'author', 'source', 'url', 'urlToImage'), partial matching on columns with longer content ('title', 'content', 'description'), and range matching on integer and date columns.

Reply:
Let me know if you think any column should be matched differently